File: src/remedy/connect/proxy.py
"""Loopback HTTP/SSE proxy used by the Connect pipe.

The phone never sees ``local_api_token``. Bearer is injected only on the
127.0.0.1 hop to the sidecar.
"""

from __future__ import annotations

import logging
from collections.abc import AsyncIterator, Mapping
from typing import Any

logger = logging.getLogger(__name__)

# Management routes refuse this so a phone cannot mint QR / retarget bind.
CONNECT_HOP_HEADER = "X-Remedy-Connect-Hop"

_HOP_BY_HOP = frozenset(
    {
        "connection",
        "keep-alive",
        "proxy-authenticate",
        "proxy-authorization",
        "te",
        "trailers",
        "transfer-encoding",
        "upgrade",
        "authorization",
        "cookie",
        "host",
        "x-remedy-token",
        "x-api-key",
        "x-forwarded-for",
        "x-forwarded-host",
        "x-forwarded-proto",
        "forwarded",
        CONNECT_HOP_HEADER.lower(),
    }
)

_STRIP_RESPONSE = frozenset(
    {
        "authorization",
        "www-authenticate",
        "proxy-authenticate",
        "x-remedy-token",
        "set-cookie",
        CONNECT_HOP_HEADER.lower(),
    }
)

_SSE_HINTS = ("text/event-stream",)


def is_sse_content_type(content_type: str) -> bool:
    low = (content_type or "").split(";", 1)[0].strip().lower()
    return any(h in low for h in _SSE_HINTS)


def _filter_request_headers(headers: Mapping[str, str]) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in headers.items():
        lk = str(key).strip().lower()
        if lk in _HOP_BY_HOP or lk.startswith("proxy-"):
            continue
        out[str(key)] = str(value)
    return out


def _filter_response_headers(headers: Mapping[str, str], *, sse: bool) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in headers.items():
        lk = str(key).strip().lower()
        if lk in {"transfer-encoding", "connection", "keep-alive"}:
            continue
        if lk in _STRIP_RESPONSE:
            continue
        if sse and lk == "content-length":
            continue
        out[str(key)] = str(value)
    return out


def encode_http_head(
    status: int,
    reason: str,
    headers: Mapping[str, str],
) -> bytes:
    lines = [f"HTTP/1.1 {int(status)} {reason or 'OK'}"]
    for key, value in headers.items():
        lines.append(f"{key}: {value}")
    lines.append("")
    lines.append("")
    return "\r\n".join(lines).encode("iso-8859-1")


def encode_http_error(status: int, reason: str, body: bytes) -> bytes:
    headers = {
        "Content-Type": "application/json",
        "Content-Length": str(len(body)),
        "Cache-Control": "no-store",
        "Connection": "keep-alive",
    }
    return encode_http_head(status, reason, headers) + body


async def iter_proxy_response(
    method: str,
    path: str,
    query: str,
    headers: Mapping[str, str],
    body: bytes,
    *,
    sidecar_port: int,
    api_key: str,
    session: Any | None = None,
) -> AsyncIterator[bytes]:
    """Yield HTTP response bytes as they arrive. SSE is not fully buffered."""
    import aiohttp

    from remedy.connect.deny import sanitize_origin_path

    port = int(sidecar_port or 7400)
    if port <= 0 or port > 65535:
        raise ValueError("sidecar port out of range")
    safe = sanitize_origin_path(path)
    if safe is None:
        body = b'{"error":"path"}'
        yield encode_http_error(400, "Bad Request", body)
        return
    q = (query or "").lstrip("?")
    url = f"http://127.0.0.1:{port}{safe}"
    if q:
        url = f"{url}?{q}"
    req_headers = _filter_request_headers(headers)
    req_headers["Host"] = f"127.0.0.1:{port}"
    req_headers[CONNECT_HOP_HEADER] = "1"
    token = str(api_key or "")
    if token:
        req_headers["Authorization"] = f"Bearer {token}"
    timeout = aiohttp.ClientTimeout(total=None, sock_connect=10, sock_read=None)
    owns_session = session is None
    client = session
    if owns_session:
        client = aiohttp.ClientSession(timeout=timeout)
    assert client is not None
    try:
        cm = client.request(
            (method or "GET").upper(),
            url,
            headers=req_headers,
            data=body if body else None,
            allow_redirects=False,
        )
        async with cm as resp:
            ctype = ""
            try:
                ctype = str(resp.headers.get("Content-Type") or "")
            except Exception:
                ctype = ""
            sse = is_sse_content_type(ctype)
            out_headers = _filter_response_headers(resp.headers, sse=sse)
            reason = getattr(resp, "reason", None) or "OK"
            status = int(getattr(resp, "status", 200) or 200)
            if sse:
                out_headers["Cache-Control"] = "no-cache"
                out_headers.setdefault("Content-Type", "text/event-stream")
                te_headers = dict(out_headers)
                te_headers["Transfer-Encoding"] = "chunked"
                te_headers.pop("Content-Length", None)
                te_headers.pop("content-length", None)
                yield encode_http_head(status, str(reason), te_headers)
                content = getattr(resp, "content", None)
                if content is not None and hasattr(content, "iter_chunked"):
                    async for chunk in content.iter_chunked(16384):
                        if not chunk:
                            continue
                        yield _chunked_block(chunk)
                elif content is not None and hasattr(content, "iter_any"):
                    async for chunk in content.iter_any():
                        if not chunk:
                            continue
                        yield _chunked_block(chunk)
                yield b"0\r\n\r\n"
                return
            # Non-SSE: stream body chunks; if Content-Length is known keep it.
            has_cl = any(k.lower() == "content-length" for k in out_headers)
            if not has_cl:
                out_headers["Transfer-Encoding"] = "chunked"
                yield encode_http_head(status, str(reason), out_headers)
                content = getattr(resp, "content", None)
                if content is not None and hasattr(content, "iter_chunked"):
                    async for chunk in content.iter_chunked(16384):
                        if chunk:
                            yield _chunked_block(chunk)
                yield b"0\r\n\r\n"
                return
            yield encode_http_head(status, str(reason), out_headers)
            content = getattr(resp, "content", None)
            if content is not None and hasattr(content, "iter_chunked"):
                async for chunk in content.iter_chunked(16384):
                    if chunk:
                        yield bytes(chunk)
            elif content is not None and hasattr(content, "read"):
                # Last resort: still read in slices, not one giant buffer.
                while True:
                    piece = await content.read(16384)
                    if not piece:
                        break
                    yield bytes(piece)
    finally:
        if owns_session:
            close = getattr(client, "close", None)
            if close is not None:
                result = close()
                if hasattr(result, "__await__"):
                    await result


def _chunked_block(chunk: bytes) -> bytes:
    return f"{len(chunk):X}\r\n".encode("ascii") + chunk + b"\r\n"

File: native/go/connect/proxy.go
package connect

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const (
	// ConnectHopHeader marks loopback hops so management routes can refuse phone minting.
	ConnectHopHeader = "X-Remedy-Connect-Hop"
)

var hopByHop = map[string]struct{}{
	"connection": {}, "keep-alive": {}, "proxy-authenticate": {}, "proxy-authorization": {},
	"te": {}, "trailers": {}, "transfer-encoding": {}, "upgrade": {},
	"authorization": {}, "cookie": {}, "host": {}, "x-remedy-token": {}, "x-api-key": {},
	"x-forwarded-for": {}, "x-forwarded-host": {}, "x-forwarded-proto": {}, "forwarded": {},
	strings.ToLower(ConnectHopHeader): {},
}

var stripResponse = map[string]struct{}{
	"authorization": {}, "www-authenticate": {}, "proxy-authenticate": {},
	"x-remedy-token": {}, "set-cookie": {}, strings.ToLower(ConnectHopHeader): {},
}

// IsSSEContentType reports text/event-stream (and friends).
func IsSSEContentType(contentType string) bool {
	low := strings.ToLower(strings.TrimSpace(strings.Split(contentType, ";")[0]))
	return strings.Contains(low, "text/event-stream")
}

func filterRequestHeaders(headers string) http.Header {
	out := make(http.Header)
	for _, line := range strings.Split(headers, "\n") {
		line = strings.TrimSpace(strings.TrimSuffix(line, "\r"))
		if line == "" {
			continue
		}
		name, value, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		lk := strings.ToLower(strings.TrimSpace(name))
		if _, drop := hopByHop[lk]; drop || strings.HasPrefix(lk, "proxy-") {
			continue
		}
		out.Add(strings.TrimSpace(name), strings.TrimSpace(value))
	}
	return out
}

func filterResponseHeaders(h http.Header, sse bool) http.Header {
	out := make(http.Header)
	for key, vals := range h {
		lk := strings.ToLower(key)
		if lk == "transfer-encoding" || lk == "connection" || lk == "keep-alive" {
			continue
		}
		if _, drop := stripResponse[lk]; drop {
			continue
		}
		if sse && lk == "content-length" {
			continue
		}
		for _, v := range vals {
			out.Add(key, v)
		}
	}
	return out
}

// EncodeHTTPHead builds an HTTP/1.1 response head (status + headers + blank line).
func EncodeHTTPHead(status int, reason string, headers http.Header) []byte {
	if reason == "" {
		reason = "OK"
	}
	var b strings.Builder
	b.WriteString(fmt.Sprintf("HTTP/1.1 %d %s\r\n", status, reason))
	for key, vals := range headers {
		for _, v := range vals {
			b.WriteString(key)
			b.WriteString(": ")
			b.WriteString(v)
			b.WriteString("\r\n")
		}
	}
	b.WriteString("\r\n")
	return []byte(b.String())
}

// EncodeHTTPError builds a JSON error response as HTTP/1.1 bytes.
func EncodeHTTPError(status int, reason string, body []byte) []byte {
	h := make(http.Header)
	h.Set("Content-Type", "application/json")
	h.Set("Content-Length", strconv.Itoa(len(body)))
	h.Set("Cache-Control", "no-store")
	h.Set("Connection", "keep-alive")
	return append(EncodeHTTPHead(status, reason, h), body...)
}

func chunkedBlock(chunk []byte) []byte {
	out := make([]byte, 0, len(chunk)+32)
	out = append(out, []byte(fmt.Sprintf("%X\r\n", len(chunk)))...)
	out = append(out, chunk...)
	out = append(out, '\r', '\n')
	return out
}

// ProxyRequest is one loopback hop to the local API.
type ProxyRequest struct {
	Method      string
	Path        string
	Query       string
	Headers     string
	Body        []byte
	SidecarPort int
	APIKey      string
}

// IterProxyResponse yields HTTP/1.1 response bytes as they arrive (SSE chunked).
func IterProxyResponse(req ProxyRequest, emit func([]byte) error) error {
	port := req.SidecarPort
	if port <= 0 {
		port = 7400
	}
	if port > 65535 {
		return fmt.Errorf("%w: sidecar port out of range", ErrSession)
	}
	safe, ok := SanitizeOriginPath(req.Path)
	if !ok {
		body := []byte(`{"error":"path"}`)
		return emit(EncodeHTTPError(400, "Bad Request", body))
	}
	q := strings.TrimPrefix(strings.TrimSpace(req.Query), "?")
	url := fmt.Sprintf("http://127.0.0.1:%d%s", port, safe)
	if q != "" {
		url = url + "?" + q
	}
	method := strings.ToUpper(strings.TrimSpace(req.Method))
	if method == "" {
		method = "GET"
	}
	var bodyReader io.Reader
	if len(req.Body) > 0 {
		bodyReader = bytes.NewReader(req.Body)
	}
	httpReq, err := http.NewRequest(method, url, bodyReader)
	if err != nil {
		return err
	}
	for k, vals := range filterRequestHeaders(req.Headers) {
		for _, v := range vals {
			httpReq.Header.Add(k, v)
		}
	}
	httpReq.Host = fmt.Sprintf("127.0.0.1:%d", port)
	httpReq.Header.Set(ConnectHopHeader, "1")
	if tok := strings.TrimSpace(req.APIKey); tok != "" {
		httpReq.Header.Set("Authorization", "Bearer "+tok)
	}
	client := &http.Client{
		Timeout: 0,
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
		Transport: &http.Transport{
			ResponseHeaderTimeout: 10 * time.Second,
		},
	}
	resp, err := client.Do(httpReq)
	if err != nil {
		body := []byte(`{"error":"pipe","reason":"pipe"}`)
		return emit(EncodeHTTPError(502, "Bad Gateway", body))
	}
	defer resp.Body.Close()

	ctype := resp.Header.Get("Content-Type")
	sse := IsSSEContentType(ctype)
	outHeaders := filterResponseHeaders(resp.Header, sse)
	reason := resp.Status
	if i := strings.IndexByte(reason, ' '); i >= 0 && i+1 < len(reason) {
		reason = reason[i+1:]
	}
	if reason == "" {
		reason = "OK"
	}
	if sse {
		outHeaders.Set("Cache-Control", "no-cache")
		if outHeaders.Get("Content-Type") == "" {
			outHeaders.Set("Content-Type", "text/event-stream")
		}
		outHeaders.Set("Transfer-Encoding", "chunked")
		outHeaders.Del("Content-Length")
		if err := emit(EncodeHTTPHead(resp.StatusCode, reason, outHeaders)); err != nil {
			return err
		}
		buf := make([]byte, 16384)
		for {
			n, rerr := resp.Body.Read(buf)
			if n > 0 {
				if err := emit(chunkedBlock(buf[:n])); err != nil {
					return err
				}
			}
			if rerr == io.EOF {
				break
			}
			if rerr != nil {
				return rerr
			}
		}
		return emit([]byte("0\r\n\r\n"))
	}

	hasCL := outHeaders.Get("Content-Length") != ""
	if !hasCL {
		outHeaders.Set("Transfer-Encoding", "chunked")
		if err := emit(EncodeHTTPHead(resp.StatusCode, reason, outHeaders)); err != nil {
			return err
		}
		buf := make([]byte, 16384)
		for {
			n, rerr := resp.Body.Read(buf)
			if n > 0 {
				if err := emit(chunkedBlock(buf[:n])); err != nil {
					return err
				}
			}
			if rerr == io.EOF {
				break
			}
			if rerr != nil {
				return rerr
			}
		}
		return emit([]byte("0\r\n\r\n"))
	}

	if err := emit(EncodeHTTPHead(resp.StatusCode, reason, outHeaders)); err != nil {
		return err
	}
	_, err = io.Copy(&emitWriter{emit: emit}, resp.Body)
	return err
}

type emitWriter struct {
	emit func([]byte) error
}

func (w *emitWriter) Write(p []byte) (int, error) {
	if len(p) == 0 {
		return 0, nil
	}
	cp := append([]byte(nil), p...)
	if err := w.emit(cp); err != nil {
		return 0, err
	}
	return len(p), nil
}
